graders/c/cgrader/cgrader.py

```diff
@@ -0,0 +1,266 @@
+#! /usr/bin/python3
+
+import os, json, subprocess, shlex, threading, re, sys
+
+CODEBASE = '/grade/student'
+DATAFILE = '/grade/data/data.json'
+SB_USER = 'sbuser'
+
+TIMEOUT_MESSAGE = \
+    '\n\nTIMEOUT! Typically this means the program took too long,' + \
+    '\nrequested more inputs than provided, or an infinite loop was found.' + \
+    '\nIf your program is reading data using scanf inside a loop, this ' + \
+    '\ncould also mean that scanf does not support the input provided ' + \
+    '\n(e.g., reading an int if the input is a double).\n'
+
+class UngradableException(Exception):
+    def __init__(self):
+        pass
+
+class CGrader:
+
+    def __init__(self):
+        with open(DATAFILE) as file:
+            self.data = json.load(file)
+
+    def run_command(self, command, input=None, sandboxed=False, timeout=None):
```

Comment:
Any objection to making `sandboxed` default to `True`, and letting users turn it off if needed?  I'd think for most cases we wouldn't need to run the student's code as root?

Reply:
My initial thought was that `test_run` calls `run_command` with `sandboxed` set to true, and most users are expected to run it from there. But you're right, if the question calls it directly, it probably makes more sense to set it to `False` by default, and change the `cgrader` internal calls (e.g., compilation) to run as root. I'll review this.